### PR TITLE
asusctl: init at 4.0.6

### PR DESCRIPTION
**Pull request by Cogitri**
###### Motivation for this change

Add new package asusctl to improve support for ASUS Laptops.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix
Comment (legendofmiracles) [inline maintainers/maintainer-list.nix:2241]: Different commit: `maintainers: add Cogitri`

File: pkgs/tools/misc/asusctl/default.nix
Comment (legendofmiracles) [inline pkgs/tools/misc/asusctl/default.nix:3]:
```suggestion
```

**Review by legendofmiracles — COMMENTED** [review]
Great work (by the looks of it, can't test it)! I tried my hands at it a while back but gave up very quickly after some complications...

**Comment by Cogitri** [comment]
Hm, for some reason the systemd system unit service isn't copied to share/systemd while the user services are

**Comment by figsoda** [comment]
should probably have more patching since asusctl tries to edit config files ([example](https://gitlab.com/asus-linux/asusctl/-/blob/ab195e1d84021fb250fe4198fcab63244df1bc69/daemon/src/ctrl_anime/config.rs#L142))

would be nice to have a module as well

File: pkgs/tools/misc/asusctl/statedir.patch
Comment (Cogitri) [reply pkgs/tools/misc/asusctl/statedir.patch:9]: Hm, StateDirectory wouldn’t create the dir for me but maybe I just typoed somewhere. Let me try that again, thanks for the heads up

Comment (Cogitri) [reply pkgs/tools/misc/asusctl/statedir.patch]: Yup that worked, not sure what I did different previously :D

File: nixos/modules/services/misc/asusctl.nix
Comment (leo60228) [inline nixos/modules/services/misc/asusctl.nix:12]: This phrasing reads a bit oddly to me.

File: pkgs/tools/misc/asusctl/default.nix
Comment (leo60228) [inline pkgs/tools/misc/asusctl/default.nix:57]:
```suggestion

```

**Comment by leo60228** [comment]
I could be wrong, but I *think* the convention is to have separate `asusctl: init at 4.0.6` and `nixos/asusctl: init` commits.

**Comment by Cogitri** [comment]
Fixed :+1:

**Review by leo60228 — APPROVED** [review]
I won't be able to test this for a few weeks, but it looks good.

**Comment by Madouura** [comment]
For the most part, it seems to be working fine.
Just a few problems, at least on my configuration.

supergfxd:
```
sudo systemctl status supergfxd
● supergfxd.service - SUPERGFX
     Loaded: loaded (/etc/systemd/system/supergfxd.service; linked; vendor preset: enabled)
     Active: active (running) since Thu 2021-12-16 10:29:19 CST; 18min ago
   Main PID: 4725 (supergfxd)
         IP: 0B in, 0B out
         IO: 0B read, 20.0K written
      Tasks: 1 (limit: 18444)
     Memory: 284.0K
        CPU: 0
     CGroup: /system.slice/supergfxd.service
             └─4725 /nix/store/ysk35xrh97hb2lrw67h2ndhkk5wzv0mg-supergfxctl-2.0.4/bin/supergfxd

Dec 16 10:29:21 tsuki supergfxd[4725]: INFO: GFX: 0000:07:00.4: Function for 0000:07:00.0
Dec 16 10:29:21 tsuki supergfxd[4725]: INFO: GFX: 0000:07:00.2: Function for 0000:07:00.0
Dec 16 10:29:21 tsuki supergfxd[4725]: INFO: GFX: 0000:07:00.0: Function for 0000:07:00.0
Dec 16 10:29:21 tsuki supergfxd[4725]: INFO: GFX: 0000:07:00.5: Function for 0000:07:00.0
Dec 16 10:29:21 tsuki supergfxd[4725]: INFO: GFX: 0000:07:00.3: Function for 0000:07:00.0
Dec 16 10:29:21 tsuki supergfxd[4725]: INFO: GFX: 0000:07:00.1: Function for 0000:07:00.0
Dec 16 10:29:21 tsuki supergfxd[4725]: INFO: GFX: 0000:01:00.0: NVIDIA graphics, setting PM to auto
Dec 16 10:29:21 tsuki supergfxd[4725]: INFO: GFX: Writing /etc/X11/xorg.conf.d/90-nvidia-primary.conf
Dec 16 10:29:21 tsuki supergfxd[4725]: INFO: GFX: Writing /etc/modprobe.d/supergfxd.conf
Dec 16 10:29:21 tsuki supergfxd[4725]: ERROR: Gfx controller: Command exec error: "modprobe" "nvidia_drm": No such file or directory (os error 2)
```

asusd:
```
● asusd.service - ASUS Notebook Control
     Loaded: loaded (/etc/systemd/system/asusd.service; linked; vendor preset: enabled)
     Active: active (running) since Thu 2021-12-16 10:43:10 CST; 4s ago
    Process: 11348 ExecStartPre=/nix/store/fvprxgcxf4px865gdjd81fbwnxcjrg41-coreutils-9.0/bin/sleep 2 (code=exited, status=0/SUCCESS)
   Main PID: 11350 (asusd)
         IP: 0B in, 0B out
         IO: 3.3M read, 12.0K written
      Tasks: 4 (limit: 18444)
     Memory: 2.4M
        CPU: 0
     CGroup: /system.slice/asusd.service
             └─11350 /nix/store/rfdn03vqb8s2kjc5llqi9hmbcrbmnrrn-asusctl-4.0.6/bin/asusd

Dec 16 10:43:10 tsuki asusd[11350]: ERROR: rog_bios_control: Path /sys/firmware/efi/efivars/AsusPostLogoSound-607005d5-3f75-4b2e-98f0-85ba66797a3e: No such fil>
Dec 16 10:43:10 tsuki asusd[11350]: INFO: Battery charge limit: 80
Dec 16 10:43:10 tsuki asusd[11350]: INFO: Device has profile control available
Dec 16 10:43:10 tsuki asusd[11350]: ERROR: AniMe control: Missing functionality: ASUS AniMe device node not found
Dec 16 10:43:10 tsuki asusd[11350]: INFO: Matched to ROG Zephyrus G15 GA503Q
Dec 16 10:43:10 tsuki asusd[11350]: WARN: led_node: Missing functionality: ASUS LED device node not found
Dec 16 10:43:10 tsuki asusd[11350]: WARN: led_node: Missing functionality: ASUS LED device node not found
Dec 16 10:43:10 tsuki asusd[11350]: WARN: led_node: Missing functionality: ASUS LED device node not found
Dec 16 10:43:10 tsuki asusd[11350]: INFO: Using device at: "/dev/hidraw5" for LED control
Dec 16 10:43:10 tsuki systemd[1]: Started ASUS Notebook Control.
```

supergfx can't seem to modprobe, for some reason.
asusd can't seem to read an efivar, which I have confirmed is there.
asusd won't start without a rebuild, and it goes without mentioning, doesn't start on boot, so I have to nixos-rebuild and then start it manually.

**Comment by Madouura** [comment]
Also, supergfxd creates ``/etc/supergfxd.conf``.
This should be controlled by the nix configuration in ``nixos/supergfxctl.nix`` preferably.

**Comment by Madouura** [comment]
From a cursory look, you should be able to more or less do what you did with asusctl and it's systemd service to supergfxctl.

**Comment by Madouura** [comment]
Something interesting I noticed, if I ``modprobe -i vfio-pci``, and then restart supergfxd, and then restart it again, it keeps failing until I ``nixos-rebuild switch``.

**Comment by Cogitri** [comment]
Thanks for the feedback @Madouura ! I've added the options to supergfxctl and fixed the paths to modprobe and rmmod. That should fix the last thing you've mentioned re: vfio-pci. I'll have to look into why asusd fails to start.

**Comment by Cogitri** [comment]
>  asusd can't seem to read an efivar, which I have confirmed is there.

I think it actually _can_ read the file, but it can't write to it since efivars is mounted as ro. I don't think we have an option in NixOS to remount it as r-w, so I'm not sure if we can do something about that.

**Comment by Madouura** [comment]
I'll test this a bit later.
You may want to [follow the contributing guidelines and add release notes](https://github.com/NixOS/nixpkgs/blob/af21d41260846fb9c9840a75e310e56dfe97d6a3/CONTRIBUTING.md#generating-2205-release-notes).

**Comment by Madouura** [comment]
I can't find much logs to go off of either, the two top things I can think of for the reason of asusd failing would be either a hard-coded path or something with starting supergfxd.
It's possible asusd may just be starting too early, or not early enough as well.

**Comment by Cogitri** [comment]
Yeah, maybe it’s starting too early. I don’t think it’s due to hard coded paths, then it wouldn’t work sometimes. I’ll try fixing it tomorrow

**Comment by Madouura** [comment]
``thread 'main' panicked at 'The directory /etc/asusd/ is missing', daemon/src/config.rs:28:33``

**Comment by Cogitri** [comment]
@Madouura Are you sure that's from `asusd` from the latest version of this PR? The logs look like this for me:

```
Jan 05 15:20:04 nixos systemd[1]: Stopped ASUS Notebook Control.
Jan 05 15:20:04 nixos systemd[1]: asusd.service: Deactivated successfully.
Jan 05 15:20:04 nixos systemd[1]: Stopping ASUS Notebook Control...
Jan 05 15:20:04 nixos asusd[1185]: INFO: rog-supported v4.0.0
Jan 05 15:20:04 nixos asusd[1185]: INFO:  rog-profiles v1.1.3
Jan 05 15:20:04 nixos asusd[1185]: INFO:      rog-dbus v4.0.2
Jan 05 15:20:04 nixos asusd[1185]: INFO:      rog-aura v1.1.0
Jan 05 15:20:04 nixos asusd[1185]: INFO:     rog-anime v1.3.0
Jan 05 15:20:04 nixos asusd[1185]: INFO:        daemon v4.0.7
Jan 05 15:20:04 nixos systemd[1]: Started ASUS Notebook Control.
```

Which would suggest that systemd is stopping it (for some reason)

**Comment by Madouura** [comment]
It should be, I'll retry it this weekend.

**Comment by Madouura** [comment]
Try replacing the contents of ``service.patch`` with
```
diff --git a/data/asusd.service b/data/asusd.service
index 9f0ee84..8b5d466 100644
--- a/data/asusd.service
+++ b/data/asusd.service
@@ -1,16 +1,18 @@
 [Unit]
 Description=ASUS Notebook Control
-StartLimitInterval=200
+StartLimitInterval=0
 StartLimitBurst=2
 Before=multi-user.target
+After=supergfxd.service
+Requires=supergfxd.service
 
 [Service]
 Environment=IS_SERVICE=1
-ExecStartPre=/bin/sleep 2
 ExecStart=/usr/bin/asusd
+StateDirectory=asusd
 Restart=on-failure
 Restart=always
-RestartSec=1
+RestartSec=4
 Type=dbus
 BusName=org.asuslinux.Daemon
 SELinuxContext=system_u:system_r:unconfined_t:s0
```

I have it working with this.

**Comment by Cogitri** [comment]
Huh weird,  I'm pretty sure I've tried something like that too - let me apply that patch, thanks a lot for the suggestion :)

**Comment by Madouura** [comment]
I couldn't find a good way to separate supergfxd and asusd unfortunately, but I fixed a couple issues in this patch.
You should just be able to ``git apply`` this right onto your nixpkgs fork.
[lol.patch.txt](https://github.com/NixOS/nixpkgs/files/7834131/lol.patch.txt)
```diff
diff --git a/pkgs/tools/misc/asusctl/default.nix b/pkgs/tools/misc/asusctl/default.nix
index b44af812..ac8dc453 100644
--- a/pkgs/tools/misc/asusctl/default.nix
+++ b/pkgs/tools/misc/asusctl/default.nix
@@ -29,8 +29,8 @@ rustPlatform.buildRustPackage rec {
     for file in data/*.service; do
       substituteInPlace $file \
         --replace /usr/bin/sleep ${coreutils}/bin/sleep \
-        --replace /bin/sleep ${coreutils}/bin/sleep \
-        --replace /bin/mkdir ${coreutils}/bin/mkdir \
+        --replace =/bin/sleep =${coreutils}/bin/sleep \
+        --replace =/bin/mkdir =${coreutils}/bin/mkdir \
         --replace /usr/bin $out/bin
     done

diff --git a/pkgs/tools/misc/asusctl/service.patch b/pkgs/tools/misc/asusctl/service.patch
index 7fd232d5..cc113539 100644
--- a/pkgs/tools/misc/asusctl/service.patch
+++ b/pkgs/tools/misc/asusctl/service.patch
@@ -1,9 +1,12 @@
 diff --git a/data/asusd.service b/data/asusd.service
-index 9f0ee84..66add89 100644
+index 9f0ee84..8b5d466 100644
 --- a/data/asusd.service
 +++ b/data/asusd.service
-@@ -3,10 +3,13 @@ Description=ASUS Notebook Control
- StartLimitInterval=200
+@@ -1,16 +1,18 @@
+ [Unit]
+ Description=ASUS Notebook Control
+-StartLimitInterval=200
++StartLimitInterval=0
  StartLimitBurst=2
  Before=multi-user.target
 +After=supergfxd.service
@@ -11,9 +14,14 @@ index 9f0ee84..66add89 100644

  [Service]
  Environment=IS_SERVICE=1
- ExecStartPre=/bin/sleep 2
-+StateDir=asusd
+-ExecStartPre=/bin/sleep 2
  ExecStart=/usr/bin/asusd
++StateDirectory=asusd
  Restart=on-failure
  Restart=always
-
+-RestartSec=1
++RestartSec=4
+ Type=dbus
+ BusName=org.asuslinux.Daemon
+ SELinuxContext=system_u:system_r:unconfined_t:s0
+
\ No newline at end of file
```

**Comment by Cogitri** [comment]
With the patch you've provided it seems to work just fine! Weird, I'm pretty sure I've tried something similar. Thanks for the help :)

**Review by Madouura — APPROVED** [review]
LGTM

**Comment by Madouura** [comment]
@Cogitri Looks like there's some merge conflicts, you'll need to rebase.
Also, I would suggest you add some release notes: https://github.com/NixOS/nixpkgs/blob/af21d41260846fb9c9840a75e310e56dfe97d6a3/CONTRIBUTING.md#generating-2205-release-notes

**Comment by Cogitri** [comment]
Done, thanks :)

**Comment by Cogitri** [comment]
Rebased against master. This has worked on my laptop for a few weeks now :)

**Comment by Madouura** [comment]
Same for me, don't see any major changes so still LGTM.

**Comment by leo60228** [comment]
@itsfarseen Clone the PR branch, and do `sudo nixos-rebuild switch -I nixpkgs=/path/to/nixpkgs`

**Comment by itsfarseen** [comment]
@leo60228, that will also update every other package right? Is there a way to somehow pick just asusctl?

**Comment by itsfarseen** [comment]
This is what I ended up doing:
```nix
let
  nixpkgs-tars = "https://github.com/NixOS/nixpkgs/archive/";
  pr_asusctl_tar = fetchTarball "${nixpkgs-tars}a4a81b6f6c27e5a964faea25b7b5cbe611f98691.tar.gz";
in
  { config, lib, pkgs, ... }:

  {
    imports =
      [ # include the results of the hardware scan.
        ./hardware-configuration.nix
        "${pr_asusctl_tar}/nixos/modules/services/misc/asusctl.nix"
        "${pr_asusctl_tar}/nixos/modules/services/misc/supergfxctl.nix"
      ];

    nixpkgs.overlays = [
      (self: super: { 
        asusctl = pkgs.callpackage "${pr_asusctl_tar}/pkgs/tools/misc/asusctl/default.nix" {};
        supergfxctl = pkgs.callpackage "${pr_asusctl_tar}/pkgs/tools/misc/supergfxctl/default.nix" {};
      })
    ];
    
    environment.systemPackages = with pkgs; [
      # .. snip ..
      asusctl
    ];

    services.asusctl.enable = true;
   
    # .. snip .. 
  }
```

**Comment by mtoohey31** [comment]
I've been testing this, and it works _almost_ perfectly, but I've been having trouble getting the Quiet profile to persist between reboots. From what I can tell, asusd loads it from `/etc/asusd/profile.conf` when it starts, but the postPatch in `asusctl/default.nix` switches things so it uses `/var/lib/asusd/profile.conf`. When I apply a profile interactively with `asusctl profile -P Quiet`, the contents of `profile.conf` change to reflect the active profile, but after a reboot it's back to Balanced every time. Any ideas on what might be causing this?

**Comment by leo60228** [comment]
Recent kernels manage those profiles now, maybe there's some kind of conflict?

**Comment by mtoohey31** [comment]
That may be the issue. asusd is definitely what's overwriting the file though, I tried setting the profile to Quiet, disabling the asusd service, then rebooting, and `profile.conf` doesn't change until I the asusd service is started again. I think what's happening is the profile isn't surviving the reboot inside the bios/whatever low level system is responsible for that, then when asusd starts it realizes the file is out of sync with what's actually happening so it updates it. I'm going to do some testing with the behaviour on my arch installation to see if the profile survives reboots without asusd running.

**Comment by mtoohey31** [comment]
Ok, sorry for cluttering up the thread, it seems I misunderstood the way asusctl does profile control. It looks like most of the work is done by power-profiles-daemon, and asusctl just provides an subcommand that does a similar thing to powerprofilectl, which is the usual interface for power-profiles-daemon. So I'm not sure whether it makes sense to include a dependency on power-profiles-daemon here, because while the profile you set doesn't persist between reboots without it, some users might not want that functionality.

For reference, the necessary configuration to get power-profiles-daemon working is:

```nix
  services.power-profiles-daemon.enable = true;
  systemd.services.power-profiles-daemon = {
    enable = true;
    wantedBy = [ "multi-user.target" ];
  };
```

Should we add a setting for the asusctl service that does this, or should we leave it up to the user to configure manually?

**Comment by Madouura** [comment]
@Cogitri

**Comment by icyphox** [comment]
Would it be better naming the service `asusd` and `supergfxd` instead? Since the `*ctl` names are for the CLIs interacting with the daemon.

**Comment by Cogitri** [comment]
Sounds like a good idea, thanks

**Comment by Cogitri** [comment]
I believe all concerns should be addressed now. For those of you using flakes, you can use my repository if you want to: https://github.com/Cogitri/cogitri-pkgs

File: pkgs/tools/misc/supergfxctl/default.nix
Comment (icyphox) [inline pkgs/tools/misc/supergfxctl/default.nix:5]: Bump to latest release [4.0.4](https://gitlab.com/asus-linux/supergfxctl/-/releases#4.0.4)?

Comment (Cogitri) [reply pkgs/tools/misc/supergfxctl/default.nix:5]: I'd rather get this merged first before I keep updating this PR over and over again to be perfectly honest

Comment (Madouura) [reply pkgs/tools/misc/supergfxctl/default.nix:5]: Yeah, let's just get this merged for now, it's gotten a bit ridiculous.
I won't be able to test this until the weekend, though.

File: pkgs/tools/misc/asusctl/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/misc/asusctl/default.nix:50]: What is that?

File: pkgs/tools/misc/supergfxctl/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/misc/supergfxctl/default.nix:35]:
```suggestion
```
Is that not working?

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:35221]: Please don't add things to the end of the list.

File: pkgs/tools/misc/supergfxctl/default.nix
Comment (Cogitri) [reply pkgs/tools/misc/supergfxctl/default.nix:35]: No, because then cargo build is run instead of make.

Comment (SuperSandro2000) [reply pkgs/tools/misc/supergfxctl/default.nix:35]: We definitely want to run our cargo build phase because it is very different from standard cargo build process. We can run anything else in post phases.
Also we already handle stripping outself and vendoring.

**Comment by mattywillo** [comment]
Gave this a quick try, something I noticed: supergfxctrl creates it's own modprobe config at `/etc/modprobe.d/supergfxd.conf`, pretty unexpected to turn on the service, run supergfxctrl once, then turn off the service and still have the nvidia kernel module blacklisted on next reboot.

Not sure the best fix, since it seems to rely on doing this for any of its features that require a reboot. Ideally the file would be cleaned up if the service is disabled but I don't believe there is a clean way to do that, maybe a patch to prevent it from writing to modprobe.d, and a config option to enable/disable the patch, so the user is aware of non-declarative changes to the system?

**Comment by Cogitri** [comment]
> Gave this a quick try, something I noticed: supergfxctrl creates it's own modprobe config at `/etc/modprobe.d/supergfxd.conf`, pretty unexpected to turn on the service, run supergfxctrl once, then turn off the service and still have the nvidia kernel module blacklisted on next reboot.
> 
> Not a showstopper but might throw some people for a loop, not sure the best fix either, since it seems to rely on doing this for any of its features that require a reboot. Ideally the file would be cleaned up if the service is disabled but I don't believe there is a clean way to do that, maybe a patch to prevent it from writing to modprobe.d, and a config option to enable/disable the patch, so the user is aware of non-declarative changes to the system?

I'd rather get this merged first before pouring more work into this PR

**Comment by mattywillo** [comment]
> I'd rather get this merged first before pouring more work into this PR

For sure, its a fairly niche situation in an already niche pkg that's only a portion of the branch, with no obviously correct solution, I agree on getting it merged as it is, just wanted to note it down

**Comment by Madouura** [comment]
cc @SuperSandro2000 
Could we get this merged please?

**Comment by Cogitri** [comment]
Done :heavy_check_mark:

File: pkgs/tools/misc/supergfxctl/default.nix
Comment (SuperSandro2000) [reply pkgs/tools/misc/supergfxctl/default.nix:35]: This still needs to be fixed properly

File: pkgs/tools/misc/asusctl/service.patch
Comment (SuperSandro2000) [inline pkgs/tools/misc/asusctl/service.patch:23]: I'd rather overwrite this within the nixos service than patch it.

**Comment by andOrlando** [comment]
Why do we want /etc/supergfxd.conf managed by nix? Would you, in some cases, want ot be able to switch graphics modes on the fly without having to rebuild your system?

**Comment by devurandom** [comment]
It would be great to have some variant of asusctl in NixOS, even if it had some quirks.

**Comment by SuperSandro2000** [comment]
> Why do we want /etc/supergfxd.conf managed by nix?

Because NixOS tries to be mostly stateless and most system wide settings are managed by it.

> Would you, in some cases, want ot be able to switch graphics modes on the fly without having to rebuild your system?

Do we? I don't know but a specialization could fix that.

**Comment by K900** [comment]
Superseded by https://github.com/NixOS/nixpkgs/pull/201081

**Pull request closed without merging**
